### PR TITLE
adding support for bond interfaces for Ubuntu OS

### PR DESCRIPTION
**Pull request by h00pz**
adding support for ubuntu bonded interfaces.  Testing with a single bond and an array of bonds

kvm-ubtu-01 bond_members='["ens224"]' bond_sub_interfaces='[{"vlanid":"215","ip":"172.16.215.10","mask":"255.255.255.0"},{"vlanid":"216","ip":"172.16.216.10","mask":"255.255.255.0"},{"vlanid":"217","ip":"172.16.217.10","mask":"255.255.255.0"},{"vlanid":"218","ip":"172.16.218.10","mask":"255.255.255.0"},{"vlanid":"219","ip":"172.16.219.10","mask":"255.255.255.0"}]'

### CODE REVIEW

**Comment by h00pz** [comment]
output of ansible run configuring the bonds on Ubuntu-16.04

TASK [bond-config : debug] **************************************************************************************************************************************************************************************************************************************************************
ok: [kvm-ubtu-01] => {
    "bond_members": [
        "ens224"
    ]
}
ok: [kvm-ubtu-02] => {
    "bond_members": [
        "ens224"
    ]
}
ok: [kvm-ubtu-04] => {
    "bond_members": [
        "ens224"
    ]
}

TASK [bond-config : debug] **************************************************************************************************************************************************************************************************************************************************************
ok: [kvm-ubtu-01] => {
    "bond_sub_interfaces": [
        {
            "ip": "172.16.215.10", 
            "mask": "255.255.255.0", 
            "vlanid": "215"
        }, 
        {
            "ip": "172.16.216.10", 
            "mask": "255.255.255.0", 
            "vlanid": "216"
        }, 
        {
            "ip": "172.16.217.10", 
            "mask": "255.255.255.0", 
            "vlanid": "217"
        }, 
        {
            "ip": "172.16.218.10", 
            "mask": "255.255.255.0", 
            "vlanid": "218"
        }, 
        {
            "ip": "172.16.219.10", 
            "mask": "255.255.255.0", 
            "vlanid": "219"
        }
    ]
}
ok: [kvm-ubtu-02] => {
    "bond_sub_interfaces": [
        {
            "ip": "172.16.215.11", 
            "mask": "255.255.255.0", 
            "vlanid": "215"
        }, 
        {
            "ip": "172.16.216.11", 
            "mask": "255.255.255.0", 
            "vlanid": "216"
        }, 
        {
            "ip": "172.16.217.11", 
            "mask": "255.255.255.0", 
            "vlanid": "217"
        }, 
        {
            "ip": "172.16.218.11", 
            "mask": "255.255.255.0", 
            "vlanid": "218"
        }, 
        {
            "ip": "172.16.219.11", 
            "mask": "255.255.255.0", 
            "vlanid": "219"
        }
    ]
}
ok: [kvm-ubtu-04] => {
    "bond_sub_interfaces": [
        {
            "ip": "172.16.215.13", 
            "mask": "255.255.255.0", 
            "vlanid": "215"
        }, 
        {
            "ip": "172.16.216.13", 
            "mask": "255.255.255.0", 
            "vlanid": "216"
        }, 
        {
            "ip": "172.16.217.13", 
            "mask": "255.255.255.0", 
            "vlanid": "217"
        }, 
        {
            "ip": "172.16.218.13", 
            "mask": "255.255.255.0", 
            "vlanid": "218"
        }, 
        {
            "ip": "172.16.219.13", 
            "mask": "255.255.255.0", 
            "vlanid": "219"
        }
    ]
}

TASK [bond-config : attach physical interface to the bond] ******************************************************************************************************************************************************************************************************************************
ok: [kvm-ubtu-01] => (item=ens224)
ok: [kvm-ubtu-04] => (item=ens224)
ok: [kvm-ubtu-02] => (item=ens224)

TASK [bond-config : attach physical interfaces to the bond] *****************************************************************************************************************************************************************************************************************************
ok: [kvm-ubtu-01] => (item={'ip': '172.16.215.10', 'mask': '255.255.255.0', 'vlanid': '215'})
ok: [kvm-ubtu-02] => (item={'ip': '172.16.215.11', 'mask': '255.255.255.0', 'vlanid': '215'})
ok: [kvm-ubtu-04] => (item={'ip': '172.16.215.13', 'mask': '255.255.255.0', 'vlanid': '215'})
ok: [kvm-ubtu-01] => (item={'ip': '172.16.216.10', 'mask': '255.255.255.0', 'vlanid': '216'})
ok: [kvm-ubtu-02] => (item={'ip': '172.16.216.11', 'mask': '255.255.255.0', 'vlanid': '216'})
ok: [kvm-ubtu-04] => (item={'ip': '172.16.216.13', 'mask': '255.255.255.0', 'vlanid': '216'})
ok: [kvm-ubtu-01] => (item={'ip': '172.16.217.10', 'mask': '255.255.255.0', 'vlanid': '217'})
ok: [kvm-ubtu-02] => (item={'ip': '172.16.217.11', 'mask': '255.255.255.0', 'vlanid': '217'})
ok: [kvm-ubtu-04] => (item={'ip': '172.16.217.13', 'mask': '255.255.255.0', 'vlanid': '217'})
ok: [kvm-ubtu-01] => (item={'ip': '172.16.218.10', 'mask': '255.255.255.0', 'vlanid': '218'})
ok: [kvm-ubtu-02] => (item={'ip': '172.16.218.11', 'mask': '255.255.255.0', 'vlanid': '218'})
ok: [kvm-ubtu-04] => (item={'ip': '172.16.218.13', 'mask': '255.255.255.0', 'vlanid': '218'})
ok: [kvm-ubtu-01] => (item={'ip': '172.16.219.10', 'mask': '255.255.255.0', 'vlanid': '219'})
ok: [kvm-ubtu-02] => (item={'ip': '172.16.219.11', 'mask': '255.255.255.0', 'vlanid': '219'})
ok: [kvm-ubtu-04] => (item={'ip': '172.16.219.13', 'mask': '255.255.255.0', 'vlanid': '219'})

TASK [bond-config : restart network service] ********************************************************************************************************************************************************************************************************************************************
changed: [kvm-ubtu-01]
changed: [kvm-ubtu-04]
changed: [kvm-ubtu-02]

File: roles/bond-config/tasks/main.yml [hunk 1]
```diff
@@ -5,29 +5,8 @@
...
-    with_items: "{{bond_sub_interfaces|list}}"
-  when: bond_sub_interfaces is defined
-
-- name: restart network service
-  service:
-    name: network
-    state: restarted
-  when: bond_members is defined or bond_sub_interfaces is defined
+- include: ubuntu.yml
+  when: ansible_distribution == 'Debian' or ansible_distribution == 'Ubuntu'
```
[the remark this answers is not in the document]
Comment (h00pz) [reply roles/bond-config/tasks/main.yml]: 10-4

File: roles/bond-config/tasks/main.yml [hunk 2]
```diff
@@ -6,7 +6,7 @@
 ##########################################################################################################
 
 - include: redhat.yml
-  when: ansible_distribution == 'CentOS' or ansible_distribution == 'Red Hat Enterprise Linux'
+  when: ansible_distribution == 'CentOS' or ansible_distribution == 'RedHat'
```
Comment (mmccarre) [inline roles/bond-config/tasks/main.yml]: This should be `when: ansible_os_family == "RedHat"`

Example https://github.com/platform9/express/blob/master/roles/neutron-prerequisites/tasks/main.yml

**Comment by h00pz** [comment]
[pmk.working.txt](https://github.com/platform9/express/files/2507780/pmk.working.txt)

Output of successful run of PMK build on Ubuntu 16.04

**Comment by h00pz** [comment]
[pmo.working.txt](https://github.com/platform9/express/files/2507796/pmo.working.txt)

Output of successful run on PMO build on 16.04

**Comment by mmccarre** [comment]
@h00pz my bond interface never came up for me after running this... am I missing something?

File: roles/bond-config/tasks/ubuntu.yml [hunk 3]
```diff
@@ -1,3 +1,43 @@
 ---
+- debug: var=bond_members
+- debug: var=bond_sub_interfaces
+
+- name: define physical interface
```
[the remark this answers is not in the document]
Comment (h00pz) [reply roles/bond-config/tasks/ubuntu.yml]: Done in commit https://github.com/platform9/express/pull/172/commits/93f4abed173b585f63bef8236a280720ca39bd05

File: pf9-express.yml [hunk 4]
```diff
@@ -30,10 +36,10 @@
...
 
 # Configure the bond
-- hosts: bond-config
-  become: true
-  roles:
-    - { role: "bond-config", when: manage_network == True }
+#- hosts: bond-config
+#  become: true
+#  roles:
+#    - { role: "bond-config", when: manage_network == True }
```
[the remark this answers is not in the document]
Comment (h00pz) [reply pf9-express.yml]: removed in commit https://github.com/platform9/express/pull/172/commits/1653f0e1aaf788f079abe10685e2beaafe88eb26

File: roles/bond-config/tasks/redhat.yml [hunk 5]
```diff
@@ -1,5 +1,33 @@
...
-- name: Create the network configuration file for the bond
-  template:
-    src: bond_redhat.j2
-    dest: "/etc/sysconfig/network-scripts/ifcfg-{{bond_ifname}}"
+##########################################################################################################
+## bond-config:
+##
+## Configure the physical and sub-interfaces behind the bond
+##########################################################################################################
+
```
[the remark this answers is not in the document]
Comment (h00pz) [reply roles/bond-config/tasks/redhat.yml]: moved them to the main yaml in commit https://github.com/platform9/express/pull/172/commits/1653f0e1aaf788f079abe10685e2beaafe88eb26

File: roles/bond-config/tasks/ubuntu.yml [hunk 6]
```diff
@@ -1,3 +1,43 @@
...
+  template:
+    src: ubuntu-ifcfg-phy-interface.j2
+    dest: "/etc/network/interfaces.d/ifcfg-{{bond_members}}.cfg"
+    mode: 0644
+  with_items: "{{bond_members}}"
+  when: bond_members is defined
+  register: bond_task_result
```
[the remark this answers is not in the document]
Comment (h00pz) [reply roles/bond-config/tasks/ubuntu.yml]: it inspects the file and reconfigures it based on the template.  We could add a section that checks if the file/interface exists and then just add a line which add it to the bond if you like.

**Pull request closed without merging**